### ISSUE TITLE
Hashable type check fails for builtins

### ISSUE THREAD

**Issue by severb** [JukkaL/mypy#679]
*I think* this should work. From what I understand, after promoting the `int`/`str`, mypy uses the usual MRO check (`TypeInfo.has_base`) in trying to determine if the resulting type is a subclass of `Hashable`. As expected, this check fails. Since `Hashable` overrides `____subclasshook__`, maybe something similar with `SupportsInt` will do the trick?

```python
def hashtest(e: Hashable) -> bool:
    return hash(e) == hash(123)

hashtest(123)  # fails
hashtest('abc') # fails
```

On a similar note, it wasn't clear for me that using the `register()` on an ABC won't work. It makes sense now thinking about it, since mypy does a static analysis but it surprised me at first. I think this is worth mentioning in the docs.

**Comment by JukkaL** [comment]
Yeah, `Hashable` isn't really supported properly right now. You should currently use type `Any` for hashable objects. We should mention this in the docs, at least.

In the longer term, mypy should support `Hashable`, but it's not a priority yet.

Registering ABCs should also be supported, but the implementation is non-trivial since subtyping should be globally consistent within a program. We may have to add an extra phase that collects all ABC registrations in the entire program before type checking any module.

**Comment by o11c** [comment]
I have a thought on how to support `Hashable`.

Instead of `Hashable` being a single trait, have *two* duck types: `DefaultHashable` and `ExplicitHashable`.

`DefaultHashable` is duck-implemented by any class that does *not* define `__hash__`, either as a function or `None`. For these purposes, `object.__hash__` is considered not to exist.

`ExplicitHashable` is duck-implemented by any class that defines `__hash__` as a function.

It is an error to set `__hash__` to `None` for a subclass of `ExplicitHashable`.

Then simply set `Hashable = Union[DefaultHashable, ExplicitHashable]`.
